### PR TITLE
Refactor TcGlobals

### PR DESCRIPTION
**Pull request by kerams**

### CODE REVIEW

File: src/Compiler/TypedTree/TcGlobals.fs [hunk 1]
```diff
@@ -878,26 +878,19 @@ type TcGlobals(
   let tref_CompilerGeneratedAttribute = findSysILTypeRef tname_CompilerGeneratedAttribute
   let tref_InternalsVisibleToAttribute = findSysILTypeRef tname_InternalsVisibleToAttribute
 
-  let mutable generatedAttribsCache = []
-  let mutable debuggerBrowsableNeverAttributeCache = None
-  let mkDebuggerNonUserCodeAttribute() = mkILCustomAttribute (findSysILTypeRef tname_DebuggerNonUserCodeAttribute, [], [], [])
-  let mkCompilerGeneratedAttribute () = mkILCustomAttribute (tref_CompilerGeneratedAttribute, [], [], [])
+  let debuggerNonUserCodeAttribute = mkILCustomAttribute (findSysILTypeRef tname_DebuggerNonUserCodeAttribute, [], [], [])
+  let compilerGeneratedAttribute = mkILCustomAttribute (tref_CompilerGeneratedAttribute, [], [], [])
```
Comment (kerams) [inline src/Compiler/TypedTree/TcGlobals.fs:882]: These are very common in compiled F# code, let's keep a single value around.

File: src/Compiler/TypedTree/TcGlobals.fs [hunk 2]
```diff
@@ -878,26 +878,19 @@ type TcGlobals(
...
-            generatedAttribsCache <- res
-            res
-       | res -> res
-    mkILCustomAttrs (attrs.AsList() @ attribs)
+      if Array.isEmpty generatedAttributes then
+          attrs
+      else
+          match attrs.AsArray() with
+          | [||] -> mkILCustomAttrsFromArray generatedAttributes
+          | attrs -> mkILCustomAttrsFromArray (Array.append attrs generatedAttributes)
```
Comment (kerams) [inline src/Compiler/TypedTree/TcGlobals.fs:893]: Minimize allocations unless necessary.

File: src/Compiler/TypedTree/TcGlobals.fs [hunk 3]
```diff
@@ -911,15 +904,9 @@ type TcGlobals(
...
         mkILCustomAttribute (findSysILTypeRef tname_DebuggerBrowsableAttribute, [typ_DebuggerBrowsableState], [ILAttribElem.Int32 n], [])
 
-  let mkDebuggerBrowsableNeverAttribute() =
-      match debuggerBrowsableNeverAttributeCache with
-      | None ->
-          let res = tref_DebuggerBrowsableAttribute 0
-          debuggerBrowsableNeverAttributeCache <- Some res
-          res
-      | Some res -> res
+  let debuggerBrowsableNeverAttribute = tref_DebuggerBrowsableAttribute 0
```
Comment (kerams) [inline src/Compiler/TypedTree/TcGlobals.fs:907]: I don't really see the point of this "cache". Even if the attribute never gets created during a compilation, it doesn't justify the mutable variable.

File: src/Compiler/TypedTree/TcGlobals.fs [hunk 4]
```diff
@@ -911,15 +904,9 @@ type TcGlobals(
...
 
-  let addNeverAttrs (attrs: ILAttributes) = mkILCustomAttrs (attrs.AsList() @ [mkDebuggerBrowsableNeverAttribute()])
+  let addNeverAttrs (attrs: ILAttributes) = mkILCustomAttrsFromArray (Array.append (attrs.AsArray()) [| debuggerBrowsableNeverAttribute |])
```
Comment (kerams) [inline src/Compiler/TypedTree/TcGlobals.fs:909]: `mkILCustomAttrs` converts the input list to array and calls `mkILCustomAttrsFromArray`. Avoid that and create the array directly instead.

**Review by T-Gro — APPROVED** [review]

**Review by 0101 — APPROVED** [review]

**Comment by psfinaki** [comment]
@kerams looks good, if you're fine with the current version feel free to resolve the comments and this will be merged

**Pull request merged**